### DISCUSSION TITLE
KobWeb with MySql!

### DISCUSSION THREAD

**Discussion by SaherAlSous** [varabyte/kobweb#487] [Q&A] [answered]
I'm learning Kobweb, and I started with a video tutorial on [youtube](https://www.youtube.com/watch?v=VVNq6yovU_0). 
The video is using Mongo, but I"m working on MySql since the environment I will be working on is using it.
I tried for 2 days to get the content of the db to show on the front end, but no luck at all.
I'm new to the whole subject and I'm sure that there is something I'm missing out but I'm not getting it.
I debugged the code using several ways, and I found out that I'm getting the data from the database and I can display it in the counsel, but when I checked the compose JvMain, I always got an empty list when the data reaches compose, despite I"m getting it from the API, the only time I managed to display something from the database was a byte array, I got numbers displayed, but when I converted back to String I got an empty list. I thought it is a conversion issue, but no longer. there is something else that I'm not seeing.
I'm connecting to the database this way:

```
object SqlConnector {

    @InitApi
    fun initializeSql(context: InitApiContext) {
    context.data.add(::request)
    }

    @JvmStatic
    fun request(qry: String): ResultSet? {
        return try {
            val myConnection = DriverManager.getConnection(
                "jdbc:mysql://localhost/dbname",
                "user",
                "password"
            )
            val statement = myConnection.createStatement(
                ResultSet.TYPE_SCROLL_INSENSITIVE,
                ResultSet.CONCUR_READ_ONLY
            )
            statement.executeQuery(
                qry  
            )
        } catch (sqlException: SQLException) {
            sqlException.printStackTrace()
            null
        }
    }
}
```
and I"m sending it to the API this way:
```
@Api(routeOverride = "signatures")
fun getContent(context: ApiContext) {
    val content = mutableListOf<Signature>()
    var resultSet: ResultSet? = null
        try {
        resultSet = SqlConnector.request("SELECT * FROM `db`.`table`")
        while (resultSet?.next() == true) {
            val id = resultSet.getInt("id")
            val name = resultSet.getString("name")
            val signature = Signature(id, name)
            content.add(signature)
        }
        context.res.setBodyText(
            Json.encodeToString(ApiListResponse.Success(data = content)) // I tried several other ways to send the bytearray 
        )
    } catch (e: SQLException) {
        e.printStackTrace()
    } finally {
        try {
            resultSet?.close()
        } catch (ioe: IOException) {
            ioe.printStackTrace()
        }
    }
}
```

I'm getting the data in the API this way( I tried other ways as well :
```
suspend fun getSignatures(): MutableList<Signature> {
    return try {
        val result = window.api.tryGet(
            apiPath = "signatures"
        )?.decodeToString()
        Json.decodeFromString<MutableList<Signature>>(result!!)
    } catch (e: Exception) {
        e.printStackTrace()
        mutableListOf(Signature(55, e.message!!))
    }
}
```
also tried it this way:
```
suspend fun getSignatures(
    onSuccess: (ApiListResponse) -> Unit,
    onError: (Exception) -> Unit
){
    try {
        val result = window.api.tryGet(
            apiPath = "signatures"
        )?.decodeToString()
        if(result != null){
            onSuccess(Json.decodeFromString(result))
        } else {
            onError(Exception("Something went wrong"))
        }
    }catch (e: Exception){
        println(e)
        onError(e)
    }
}
```

in the compose side I'm getting the data this way:
```
@Page
@Composable
fun HomePage() {
    val response by remember { mutableStateOf<MutableList<Signature>>(mutableListOf()) }

    LaunchedEffect(Unit) {
        // Fetch data asynchronously and update the response state
        response.addAll(getSignatures())
    }

    Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
        // Render UI based on the response state
        if (response.isEmpty()) {
            Text("List is empty")
        } else {
            response.forEach {
                Text(it.name)
            }
        }
    }
}
```
also got it this way before
```
  var response by remember {  mutableStateOf<ApiListResponse>(ApiListResponse.Loading) }

    LaunchedEffect(Unit){
        getSignatures(
            onSuccess = {
                if(it is ApiListResponse.Success){
                    response = it
                }
            },
            onError = {
                println("Error")
                response = ApiListResponse.Error(it.message.toString())
            }
        )
    }
```
but all what I get is empty list!

**Comment by bitspittle** [comment]
First question for you is if you're using [server logs](https://github.com/varabyte/kobweb#kobweb-server-logs) or [remote debugging](https://github.com/varabyte/kobweb#debugging-the-backend) to check if the problem is with MySQL (or its setup)

**Reply by SaherAlSous** [reply]
I installed MySql server locally.
I created a main function in the JvmMain package.

```
fun main() {
    getContent()
}
```

the aim is to isolate the api from the site, and change a little in the function to be like this

```
//@Api(routeOverride = "signatures")
fun getContent(/*context: ApiContext*/) {
    val content = mutableListOf<Signature>()
    var resultSet: ResultSet? = null
        try {
        resultSet = SqlConnector.request("SELECT * FROM `kobweb`.`signatures`")
        while (resultSet?.next() == true) {
            val id = resultSet.getInt("id")
            val name = resultSet.getString("name")
            // Serialize data
            val signature = Signature(id, name)
            content.add(signature)
        }
//        context.res.setBodyText(
//            Json.encodeToString(ApiListResponse.Success(data = content))
//        )
            println(content)  // <-- checking if the content is printed in the log
    } catch (e: SQLException) {
        e.printStackTrace()
    } finally {
        try {
            resultSet?.close()
        } catch (ioe: IOException) {
            ioe.printStackTrace()
        }
    }
}
```
and I got the content printed correctly in the log

```
6:18:03 PM: Executing 'jvmRun -DmainClass=com.kobweb.demo.MainKt --quiet --stacktrace'...

[Signature(id=1, name=Nahud), Signature(id=2, name=Saher)]  <--- here
6:18:05 PM: Execution finished 'jvmRun -DmainClass=com.kobweb.demo.MainKt --quiet --stacktrace'.
```

**Reply by bitspittle** [reply]
Ok so good news it seems like MySQL is working?

Can you log the json object that gets set as the body of the response? Then use browser dev tools on the client to make sure that value is getting sent back.

You can also check `kobweb create examples/todo` to see a working example of transferring json objects from server to client.

**Reply by SaherAlSous** [reply]
yes MySQL it is working...I used it in other projects as well!
never used the browser dev tools before, will check it and get back to you!

**Reply by bitspittle** [reply]
Good luck!

Browser dev tools are very useful and even just learning the basics will help so much. It's a good skill!

Check out https://developer.chrome.com/docs/devtools/network in this case

**Reply by SaherAlSous** [reply]
I Checked the DevTools, I found that I'm getting an empty list! 
> data: []

I changed the code also to check for that

```
suspend fun getSignatures(): MutableList<Signature> {
    return try {
        val result = window.api.tryGet(
            apiPath = "signatures"
        )?.decodeToString()
       if (result == null) mutableListOf(Signature(555, "Null list"))
        else {
           val content = Json.decodeFromString<MutableList<Signature>>(result)
           if (content.isEmpty()) mutableListOf(Signature(999, "Empty List"))
           else content
        }
    } catch (e: Exception) {
        e.printStackTrace()
        mutableListOf(Signature(55, e.message!!))
    }
}
```

I went back to the JVM package and printed the result that I get from MySQL, I found it correctly fetched.
it seems there might be something missing either in the dependencies or plugins, because if this is the correct way to send the data through the API
```
context.res.setBodyText(
            Json.encodeToString(content)
        )
```
and I'm getting it correctly as you see in the code above. then there is something in the backend of the whole project
here are the dependencies and plugins I have, anything missing?
```
[libraries]
kobweb-api = { module = "com.varabyte.kobweb:kobweb-api", version.ref = "kobweb" }
kobweb-core = { module = "com.varabyte.kobweb:kobweb-core ", version.ref = "kobweb" }
kobweb-silk = { module = "com.varabyte.kobweb:kobweb-silk", version.ref = "kobweb" }
kobweb-worker = { module = "com.varabyte.kobweb:kobweb-worker", version.ref = "kobweb" }
kobwebx-markdown = { module = "com.varabyte.kobwebx:kobwebx-markdown", version.ref = "kobweb" }
silk-foundation = { module = "com.varabyte.kobweb:silk-foundation", version.ref = "kobweb" }
silk-icons-fa = { module = "com.varabyte.kobwebx:silk-icons-fa", version.ref = "kobweb" }
silk-icons-mdi = { module = "com.varabyte.kobwebx:silk-icons-mdi", version.ref = "kobweb" }
kotlinx-serialization-json = { module = "org.jetbrains.kotlinx:kotlinx-serialization-json", version.ref = "serialization"}
mysql-java-connector = { module = "mysql:mysql-connector-java", version.ref = "mysql-connector" }
kotlinx-coroutine = { module = "org.jetbrains.kotlinx:kotlinx-coroutines-core", version.ref = "coroutine" }

[plugins]
jetbrains-compose = { id = "org.jetbrains.compose", version.ref = "jetbrains-compose" }
kobweb-application = { id = "com.varabyte.kobweb.application", version.ref = "kobweb" }
kobweb-library = { id = "com.varabyte.kobweb.library", version.ref = "kobweb" }
kobweb-worker = { id = "com.varabyte.kobweb.worker", version.ref = "kobweb" }
kobwebx-markdown = { id = "com.varabyte.kobwebx.markdown", version.ref = "kobweb" }
kotlin-multiplatform = { id = "org.jetbrains.kotlin.multiplatform", version.ref = "kotlin" }
kotlin-serialization = { id = "org.jetbrains.kotlin.plugin.serialization", version.ref = "kotlin" }
```

**Reply by SaherAlSous** [reply]
In case all the dependencies are correct I shall remove the whole db and install it again. maybe there is something wrong in it that I didn't notice till now!

**Reply by SaherAlSous** [reply]
okey... I think I located the problem.
I opened MySQL workbench and checked the activity to the DB.
each time I run the query from the JVM main function, I can see a select count for it.
but each time I request the query from the API, nothing happens in the database. no SELECT query is sent to it.
it seems that the API does not initialize the database connection.
do you know how it can be initialized properly?

**Reply by bitspittle** [reply]
I'm afraid I can't help you, I haven't used MySQL. You might consider joining the Discord or Kotlin slack and asking around there if anyone has any open source MySQL projects you can look at.

Otherwise, I think you have all the tools in place now to debug on your own quite effectively. I would use logging during server startup (in my `@InitApi` method) to verify that the DB is being created correctly -- I would create a dummy table, put some dummy data inside it, and then check that I can query it.

BTW, your `@InitApi` method looks weird to me. You should be using it to create instances of services, but you're using it to register a function? Please review https://github.com/varabyte/kobweb#initapi-methods-and-initializing-services if you haven't read it already.

There's nothing really *that* magical about a Kobweb server. It's just a basic Kotlin app, except I'm the one who controls the main function (and I set up a ktor server for you).

You might want to create a simple, basic Kotlin application from scratch, where all it does is connect to your DB and query a dummy value from it. I think doing this will help you understand the basics of MySQL without getting distracted by Kobweb stuff. Once you have your simple code working, you can then move it over into your Kobweb project.

**Reply by SaherAlSous** [reply]
I will check that.
thank you for your help!

**Reply by SaherAlSous** [reply]
I spent some time working on something else, and today I got back to this issue. 
The good news is that I found a way.
couldn't get the whole list for now, but only a String to the front end from MySql

```
class SqlConnection {

    init {
        initDB()
    }

    object Signatures : Table("signatures") {
        val id = integer("id")
        val name = varchar("name", length = 255)
    }

    private fun initDB() {
        val url = "jdbc:mysql://user:password@localhost:3306/db_table?useUnicode=true&serverTimezone=UTC"
        val driver = "com.mysql.cj.jdbc.Driver"
        Database.connect(url, driver)
    }


    fun selectAll(): List<Signature> {
        val sigList = mutableListOf<Signature>()
        transaction {
            val result = Signatures.selectAll()
            for (sig in result) sigList.add(
                Signature(
                    id = sig[Signatures.id],
                    name = sig[Signatures.name]
                )
            )
        }
        return sigList
    }

}
```
still when I send the whole list to the API I get a null result at the other end.
will work on it.
Thank you for your responses!

**Comment by SaherAlSous** [comment, marked as answer]
that is how I solved it!

```
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.transactions.transaction


class SqlConnection {

    init {
        initDB()
    }

    object Signatures : Table("signatures") {
        val id = integer("id")
        val name = varchar("name", length = 255)
    }

    private fun initDB() {
        val url = "jdbc:mysql://user:password@localhost:3306/db_table?useUnicode=true&serverTimezone=UTC"
        val driver = "com.mysql.cj.jdbc.Driver"
        Database.connect(url, driver)
    }


    fun selectAll(): List<Signature> {
        val sigList = mutableListOf<Signature>()
        transaction {
            val result = Signatures.selectAll()
            for (sig in result) sigList.add(
                Signature(
                    id = sig[Signatures.id],
                    name = sig[Signatures.name]
                )
            )
        }
        return sigList
    }

}
```

**Comment by bitspittle** [comment]
Appreciate the update and glad you are working. Thanks for sharing your code!

**Reply by SaherAlSous** [reply]
welcome.
I used exposed core and jdbc only.
good luck!
